### PR TITLE
🐛 Avoids service shutdown due to swarm networking errors

### PR DESCRIPTION
**Pull request by GitHK**
### What was happening?

Usually the `director-v2` and the `dynamic-sidecars` run on different nodes inside the cluster.
Sometimes, usually due to network load, network connectivity can get unstable for smaller periods of time.

During such periods all requests to inspect the dynamic-sidecar could fail causing the dynamic-sidecar to be marked as failed and removed.
The effect is that the service's interface would not be working as expected and a pop-up would come out asking for authentication.

### How was it addressed

If network connectivity is bad there is a window of time defined by `DYNAMIC_SIDECAR_NETWORK_ISSUES_TOLERANCE_S` in which the dynamic-sidecar should respond once again:
- If no correct response is obtained, the services will be marked as failed and removed.
- If the services reply again, the service will remain in place

### Improvements
- logs have been added to warn if network is degrading, each time the observation cycle is triggered a message will be displayed while the error persists.

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-issues/issues/726
- https://github.com/ITISFoundation/osparc-issues/issues/725

## Checklist

- [x] Unit tests for the changes exist
- [ ] Runs in the swarm

### CODE REVIEW

**Review by pcrespov — CHANGES_REQUESTED** [review]
The way I understand it: you have a well defined exception raised (due to network problems) and you want to ignore consecutive failures within a given time period (what you call TOLERANCE)...

Although the idea you are implementing is simple but i find *very difficult to understand* in your code mainly because:
1. the choice of variable names (very long redundant names)
2. the lack of concept encapsulation. i.e. the pieces of the logic are dispersed througout the service code!

I think the whole idea could be encapsulated in a simple error handler class that can then be used to handle this exception the way you want. An instance of this handler has be to then stored in every sidecar state object.  I wrote this and tests in  https://github.com/pcrespov/sandbox-python/blob/4db1ae54f5ef18014702407228bede431cfd16ca/py-exceptions-error-handling/test_exception_handling_with_state.py

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py [hunk 1]
```diff
@@ -244,9 +248,46 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
...
+                    # tolerance was surpassed, in that case raise an error
+                    time_since_first_error = (
+                        utc_now
+                        - scheduler_data.dynamic_sidecar.datetime_dy_sidecar_became_unreachable
+                    )
+                    if (
+                        time_since_first_error.total_seconds()
+                        > dynamic_sidecar_settings.DYNAMIC_SIDECAR_NETWORK_ISSUES_TOLERANCE_S
+                    ):
+                        logger.error(
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py:277]: I agree, in the refactor this was changed.

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_events.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    minimal_app: FastAPI,
+    scheduler_data: SchedulerData,
+    caplog_info_level: LogCaptureFixture,
+):
+    with pytest.raises(BaseClientHTTPError):
+        for _ in range(REPEAT_COUNT):
+            await events.GetStatus.action(minimal_app, scheduler_data)
+            await asyncio.sleep(SLEEP_BETWEEN_CALLS)
+
+    assert caplog_info_level.text.count("Could not contact") > 1
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/tests/unit/test_modules_dynamic_sidecar_events.py:98]: I get the point and I remember seeing it somewhere but I don't remember how it works.

**Review by pcrespov — APPROVED** [review]

**Review by elisabettai — COMMENTED** [review]
Thank you so much for this! Really looking forward to an improved user experience.

**Review by elisabettai — APPROVED** [review]

**Pull request merged**